CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,288 @@
+    std::wstring exePathStr(exePath);
+    exePathStr = L"\"" + exePathStr + L"\" -ToastActivated";
+    exePath = new wchar_t[exePathStr.length() + 1];
+    wcscpy(exePath, exePathStr.c_str());
+
+    // We don't need to worry about overflow here as ::GetModuleFileName won't
+    // return anything bigger than the max file system path (much fewer than max of DWORD).
+    DWORD dataSize = static_cast<DWORD>((::wcslen(exePath) + 1) * sizeof(WCHAR));
+
+    // Register the EXE for the COM server
+    return HRESULT_FROM_WIN32(::RegSetKeyValue(
+        HKEY_CURRENT_USER,
+        subKey,
+        nullptr,
+        REG_SZ,
+        reinterpret_cast<const BYTE*>(exePath),
+        dataSize));
+}
+
+HRESULT DesktopNotificationManagerCompat::CreateToastNotifier(IToastNotifier **notifier)
+{
+    EnsureRegistered();
+
+    ComPtr<IToastNotificationManagerStatics> toastStatics;
+    HRESULT hr = Windows::Foundation::GetActivationFactory(
+        HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotificationManager).Get(),
+        &toastStatics);
+    if (SUCCEEDED(hr))
+    {
+        if (s_aumid != nullptr)
+        {
+            hr = toastStatics->CreateToastNotifierWithId(HStringReference(s_aumid).Get(), notifier);
+        }
+
+        else
+        {
+            hr = toastStatics->CreateToastNotifier(notifier);
+        }
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::CreateToastNotification(ABI::Windows::Data::Xml::Dom::IXmlDocument *content, IToastNotification **notification)
+{
+    ComPtr<IToastNotificationFactory> factory;
+    auto hr = Windows::Foundation::GetActivationFactory(
+        HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotification).Get(),
+        &factory);
+    if (SUCCEEDED(hr))
+    {
+        hr = factory->CreateToastNotification(content, notification);
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::get_History(DesktopNotificationHistoryCompat *value)
+{
+    EnsureRegistered();
+
+    ComPtr<IToastNotificationManagerStatics> toastStatics;
+    HRESULT hr = Windows::Foundation::GetActivationFactory(
+        HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotificationManager).Get(),
+        &toastStatics);
+    if (SUCCEEDED(hr))
+    {
+        ComPtr<IToastNotificationManagerStatics2> toastStatics2;
+        hr = toastStatics.As(&toastStatics2);
+        if (SUCCEEDED(hr))
+        {
+            ComPtr<IToastNotificationHistory> history;
+            hr = toastStatics2->get_History(&history);
+            if (SUCCEEDED(hr))
+            {
+                *value = DesktopNotificationHistoryCompat(s_aumid, history);
+            }
+        }
+    }
+
+    return S_OK;
+}
+
+bool DesktopNotificationManagerCompat::CanUseHttpImages()
+{
+    return IsRunningAsUwp();
+}
+
+void DesktopNotificationManagerCompat::EnsureRegistered()
+{
+    // If not registered AUMID yet
+    if (!s_registeredAumidAndComServer)
```

Comment:
http://en.cppreference.com/w/cpp/thread/call_once use this for thread safe call once stuff.